base/threadingconstructs.jl

```diff
@@ -18,39 +18,52 @@ on `threadid()`.
 """
 nthreads() = Int(unsafe_load(cglobal(:jl_n_threads, Cint)))
 
-function _threadsfor(iter,lbody)
+function _threadsfor(iters,lbody)
     fun = gensym("_threadsfor")
-    lidx = iter.args[1]         # index
-    range = iter.args[2]
+    if iters.head == :(=)
+        lidxs = Expr(:tuple, iters.args[1])
+        ranges = Expr(:tuple, iters.args[2])
+    else
+        # note the `reverse` to put the last "lidx" first (since it varies
+        # first) to match array ordering and make the code below clearer
+        lidxs = Expr(:tuple, (iter.args[1] for iter in reverse(iters.args))...)
+        ranges = Expr(:tuple, (iter.args[2] for iter in reverse(iters.args))...)
+    end
+    ndim = length(lidxs.args)
+    # recursively get all symbols in the iteration variables, which might
+    # look like ((i,j),k,..), so we can declare them local
+    allsyms(ex) = (typeof(ex) == Symbol) ? [ex] : vcat(map(allsyms,ex.args)...)
+    localvars = allsyms(lidxs)
     quote
         function $fun()
+            # get dimensions of loops, N1 x N2 x N3....
+            rs = $(esc(ranges))
+            dims = map(length,rs)
+
+            # split linear index of work, 1:N1*N2*N3*..., evenly among threads
             tid = threadid()
-            r = $(esc(range))
-            # divide loop iterations among threads
-            len, rem = divrem(length(r), nthreads())
-            # not enough iterations for all the threads?
+            q, r = divrem(*(dims...), nthreads())
+            len = q + (tid <= r ? 1 : 0)
             if len == 0
-                if tid > rem
-                    return
-                end
-                len, rem = 1, 0
+                return
             end
-            # compute this thread's iterations
-            f = 1 + ((tid-1) * len)
-            l = f + len - 1
-            # distribute remaining iterations evenly
-            if rem > 0
-                if tid <= rem
-                    f = f + (tid-1)
-                    l = l + tid
-                else
-                    f = f + rem
-                    l = l + rem
+            start = q*(tid-1) + min(tid-1,r) + 1
+
+            # get N-d index (i1,i2,i3...) from linear index (and move back one
+            # since the first thing we do in the for loop is increment)
+            i_lidxs = [ind2sub(dims,start-1)...]
+
+            for i=1:len
+                # increment and carry N-d index (i1,i2,i3,...)
+                i_lidxs[1] += 1
+                for j=1:$ndim
+                    if i_lidxs[j]>dims[j]
```

Comment:
Thanks, yea, that's a much better. I will play around with benchmarking to see how this compares to both the original code and un@thread'ed.